docs/design/osd-topology-spread-constraints.md

```diff
@@ -0,0 +1,104 @@
+   soon as this happens, the Kubernetes scheduler acts as if the AntiAffinity
+   doesn't exist, which means that an even distribution of Pods can not be
+   guaranteed.
+2. OSD Pods are locked to only three failure domains, even when more than three
+   exist.
+
+### Goals
+
+Provide an even-as-possible distribution of OSD Pods across all available
+failure domains based on Pod count as opposed to the default of resource
+availability.
+
+### Non-Goals
+
+Always maintain an even distribution of OSD Pods. OSDs can scale at a rate that
+does not match the number of Nodes in a given failure domain, as such there will
+sometimes be uneven distribution of OSD Pods across Nodes in a given failure domain.
+
+## Proposal
+
+Implement the use of Topology Spread Constraints for Ceph OSD Pods. Include the
+ability to specify them in the `StorageCluster.StorageDeviceSet` spec so they
+can be propogated to the resulting `CephCluster`'s SCDSes, providing sane
+defaults if they are not specified.
+
+### Risks and Mitigations
+
+There is an issue in upgrade on the Rook-Ceph side that needs to be addressed.
+The issue is that we want to change the generation of SCDSes from multiple sets
+per StorageDeviceSet to one. This would cause a change in the name generation
+for OSD Pods, which would be a problem in upgrade scenarios going from three
+SCDSes to one.
+
+The initial reason for having multiple StorageClassDeviceSets was to force an
+even distribution of OSD Pods across failure domains (initially, AWS
+Availability Zones (AZs)), since we could not achieve this with Node and Pod
+affinities. Topology Spread Constraints removes the need for this hack.
+
+We will mitigate this by making an initial patch to partially implement
+Topology Spread Constraints in ocs-operator while still maintaining backwards
+compatibility with Rook-Ceph until the upgrade issue can be addressed.
+
+
+## Design Details
+
+The changes will come in two phases:
+
+**Phase 1**
+
+Change the default Placement of StorageDeviceSets to only use Topology Spread
+Constraints, but maintain the split to three StorageClassDeviceSets in the
+generated CephCluster. This preserves the naming scheme for the OSD Pods that
+Rook-Ceph currently implements.
+
+Since we will no longer need to explicitly distinguish between exact failure
+domains, the constraints themselves would be identical across all SCDSes. This
+means that all Pods will be schedulable across all failure domains, regardless
+of which SCDS it belongs to.
+
+For upgrade scenarios, existing OSD Pods would still remain with the old
+Placement configurations. However, because all OSD Pods have appropriate app
+labels to identify them, they would still influence any new OSD Pods created
+with Topology Spread Constraints, allowing the new OSD Pods to be scheduled
+in a more even pattern across failure domains.
+
+**Phase 2**
+
+The currently proposed solution for the upgrade scenario is to enhance
```

Comment:
The only downside of this phase never being implemented is that OCS would indefinitely have 3 device sets instead of 1, basically just an implementation detail. Not sure the complexity of the upgrade will be worth the change in rook since everything is already working as required for phase 1.

Reply:
And I don't want OCS to indefinitely generate three device sets. :) It will become a vestigial implementation detail that we will have to explain until the end of time (or a v2 API, whichever comes first). Also, I vaguely remember looking at the Rook-Ceph code and thinking it wouldn't be all that complex, and even that it would be a cleaner way to handle OSD detection overall.

Reply:
The implicit downside of OCS always having 3 device sets is that OCS can not use more than 3 failure domains. E.g. if you have more than 3 zones (AZs), then still only three would get used. This is s serious limitation that we want to get rid of. We *might* be able to implement a more complicated algorithm in ocs-operator, but the approach to do this in rook with one SCDS seems more robust and correct to me.

It is also important to remember that we originally introduced the splitting of the SDSs into 3 SCDSs only to work around for the absence of the TSCs, hoping/planning to get rid of this splitting again once TSCs are available.